agent/engine/containermetadata/write_metadata.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/cihub/seelog"
+)
+
+const (
+	metadataJoinSuffix = "metadata"
+	metadataFile       = "metadata.json"
+	readOnlyPerm       = 0644
+)
+
+// getTaskIDfromArn parses a task Arn and produces the task ID
+// A task Arn has format arn:aws:ecs:[region]:[account-id]:task/[task-id]
+// For a correctly formatted Arn we split it over ":" into 6 parts, the last part
+// containing the task-id which we extract by splitting it by "/".
+func getTaskIDfromArn(taskarn string) string {
+	colonSplitArn := strings.SplitN(taskarn, ":", 6)
+	// Incorrectly formatted Arn (Should not happen)
+	if len(colonSplitArn) < 6 {
+		seelog.Errorf("Failed to parse task Arn: invalid TaskArn %s", taskarn)
+		return ""
+	}
+	arnTaskPartSplit := strings.SplitN(colonSplitArn[5], "/", 2)
+	// Incorrectly formatted Arn (Should not happen)
+	if len(arnTaskPartSplit) < 2 {
+		seelog.Errorf("Failed to parse task Arn: invalid TaskArn %s", taskarn)
+		return ""
+	}
+	return arnTaskPartSplit[1]
+}
+
+// getMetadataFilePath gives the metadata file path for any agent-managed container
+func getMetadataFilePath(task *api.Task, container *api.Container, dataDir string) (string, error) {
+	taskID := getTaskIDfromArn(task.Arn)
+	// Empty task ID indicates malformed Arn (Should not happen)
+	if taskID == "" {
+		err := fmt.Errorf("Failed to form metdata file path: invalid task Arn")
+		return "", err
+	}
+	return filepath.Join(dataDir, metadataJoinSuffix, taskID, container.Name), nil
+}
+
+// mdFileExist checks if metadata file exists or not
+func mdFileExist(task *api.Task, container *api.Container, dataDir string) bool {
+	mdFileDir, err := getMetadataFilePath(task, container, dataDir)
+	// Case when file path is invalid (Due to malformed task Arn)
+	if err != nil {
+		seelog.Errorf("Failed to find metadata file: %v", err)
+		return false
+	}
+
+	mdFilePath := filepath.Join(mdFileDir, metadataFile)
+	if _, err = os.Stat(mdFilePath); err != nil {
+		if !os.IsNotExist(err) {
+			// We should specifically log any error besides "IsNotExist"
+			seelog.Errorf("Failed to find metadata file: %v", err)
+		}
+		return false
+	}
+	return true
+}
+
+// writeToMetadata puts the metadata into JSON format and writes into
+// the metadata file
+func (md *Metadata) writeToMetadataFile(task *api.Task, container *api.Container, dataDir string) error {
+	data, err := json.MarshalIndent(md, "", "\t")
+	if err != nil {
+		return err
+	}
+	mdFileDir, err := getMetadataFilePath(task, container, dataDir)
+	// Boundary case if file path is bad (Such as if task arn is incorrectly formatted)
+	if err != nil {
+		return fmt.Errorf("Failed to write to metadata file: %v", err)
+	}
+	mdFilePath := filepath.Join(mdFileDir, metadataFile)
+
+	return ioutil.WriteFile(mdFilePath, data, readOnlyPerm)
+}
+
+// getMetdataFileUpdateTime gets the last update time of the metadata file
+func getMetadataFileUpdateTime(task *api.Task, container *api.Container, dataDir string) (time.Time, error) {
```

Comment:
Why do we need to query the filesystem to find this?  Shouldn't we just store this ourself?